File: openrouter.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type ModelMessage struct {
	Role    string `json:"role"`
	Content any    `json:"content"`
}

type OpenRouterClient struct {
	APIKey string
	Model  string
	Client *http.Client
}

func (o *OpenRouterClient) Complete(ctx context.Context, messages []ModelMessage) (string, error) {
	body, err := json.Marshal(map[string]any{"model": o.Model, "messages": messages})
	if err != nil {
		return "", err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, "https://openrouter.ai/api/v1/chat/completions", bytes.NewReader(body))
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bearer "+o.APIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Title", "Beeper Thread Agent")
	client := o.Client
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(io.LimitReader(resp.Body, 2<<20))
	if err != nil {
		return "", err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("OpenRouter %d: %s", resp.StatusCode, strings.TrimSpace(string(data)))
	}
	var result struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		return "", err
	}
	if len(result.Choices) == 0 || strings.TrimSpace(result.Choices[0].Message.Content) == "" {
		return "", fmt.Errorf("OpenRouter returned an empty response")
	}
	return strings.TrimSpace(result.Choices[0].Message.Content), nil
}

File: src/openrouter.ts
export type ModelContent = string | Array<
  | { type: "text"; text: string }
  | { type: "image_url"; image_url: { url: string } }
>;

type ChatMessage = { role: "system" | "user" | "assistant"; content: ModelContent };

export class OpenRouterClient {
  constructor(private apiKey: string, private model: string) {}

  async complete(messages: ChatMessage[]): Promise<string> {
    const response = await fetch("https://openrouter.ai/api/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
        "X-Title": "Beeper Thread Agent",
      },
      body: JSON.stringify({ model: this.model, messages }),
    });
    if (!response.ok) throw new Error(`OpenRouter ${response.status}: ${await response.text()}`);
    const data = await response.json() as { choices?: Array<{ message?: { content?: string } }> };
    const text = data.choices?.[0]?.message?.content?.trim();
    if (!text) throw new Error("OpenRouter returned an empty response");
    return text;
  }
}
